pkg/service/backup/restore_worker_data.go

```diff
@@ -0,0 +1,667 @@
+		return nil, errors.Wrap(err, "validate free disk space")
+	}
+
+	takenIDs := w.chooseIDsForBatch(ctx, h.Shards, target.BatchSize)
+	if ctx.Err() != nil {
+		w.returnBatchToPool(takenIDs)
+		return nil, ctx.Err()
+	}
+	if takenIDs == nil {
+		return nil, nil //nolint: nilnil
+	}
+
+	w.Logger.Info(ctx, "Created new batch",
+		"host", h.Host,
+		"sstable_id", takenIDs,
+	)
+
+	jobID, err := w.startDownload(ctx, h.Host, dstDir, srcDir, takenIDs)
+	if err != nil {
+		w.returnBatchToPool(takenIDs)
+		return nil, err
+	}
+
+	pr := &RestoreRunProgress{
+		ClusterID:    run.ClusterID,
+		TaskID:       run.TaskID,
+		RunID:        run.ID,
+		ManifestPath: run.ManifestPath,
+		Keyspace:     run.Keyspace,
+		Table:        run.Table,
+		Host:         h.Host,
+		AgentJobID:   jobID,
+		SSTableID:    takenIDs,
+	}
+
+	w.insertRunProgress(ctx, pr)
+
+	return pr, nil
+}
+
+// startDownload creates rclone job responsible for downloading SSTables with given IDs.
+func (w *restoreWorker) startDownload(ctx context.Context, host, dstDir, srcDir string, sstableIDs []string) (int64, error) {
+	batch := w.batchFromIDs(sstableIDs)
+
+	jobID, err := w.Client.RcloneCopyPaths(ctx, host, dstDir, srcDir, batch)
+	if err != nil {
+		return 0, errors.Wrap(err, "download batch to upload dir")
+	}
+
+	w.Logger.Info(ctx, "Started downloading files",
+		"host", host,
+		"job_id", jobID,
+		"batch", batch,
+	)
+
+	return jobID, nil
+}
+
+// validateRunningJob checks if rclone job described in RestoreRunProgress is in running or success state.
+func (w *restoreWorker) validateRunningJob(ctx context.Context, pr *RestoreRunProgress) (bool, error) {
+	job, err := w.Client.RcloneJobProgress(ctx, pr.Host, pr.AgentJobID, w.Config.LongPollingTimeoutSeconds)
+	if err != nil {
+		return false, err
+	}
+
+	if job.Status == string(scyllaclient.JobSuccess) || job.Status == string(scyllaclient.JobRunning) {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// chooseIDsForBatch returns slice of IDs of SSTables that the batch consists of.
+func (w *restoreWorker) chooseIDsForBatch(ctx context.Context, shards uint, size int) []string {
+	var (
+		batchSize = size * int(shards)
+		takenIDs  []string
+	)
+
+	// All restore hosts are trying to get IDs for batch from the pool.
+	// Pool is closed after the whole table has been restored.
+
+	// Take at most batchSize IDs
+	for i := 0; i < batchSize; i++ {
+		select {
+		case id, ok := <-w.bundleIDPool:
+			if !ok {
+				return takenIDs
+			}
+			takenIDs = append(takenIDs, id)
+			if ctx.Err() != nil {
+				return takenIDs
+			}
+		default:
+			// Don't wait for more IDs if the pool is empty
+			// and host already has something to restore.
+			if len(takenIDs) > 0 {
+				return takenIDs
+			}
+			// Here host hasn't taken any IDs and pool is empty,
+			// so it waits for the whole table to be restored or
+			// for IDs that might return to the pool in case of error on the other hosts.
```

Comment:
You wrote that 
``` or for IDs that might return to the pool in case of error on the other hosts```

Is it a must to pickup from the pool IDs that went back to the pool in case of an error ?
I guess not. Restore may end with error and we accept it (as long as it can be fixed with another run).

Let's not do it. It complicates the code much. Besides, we still don't guarantee that all IDs that went back to the pool will be processed again.

Reply:
But isn't it useful to have this kind of fallback implemented (especially that it's already there)? As far as I'm concerned, restore can be a really expensive operation (downloading data from backup location and using load and stream mechanism), so maybe it makes sense to try to execute it even if one node is going through some unexpected errors?